AngularJSAuthentication.API/Providers/SimpleRefreshTokenProvider.cs

```diff
@@ -1,86 +1,122 @@
-               
-                var token = new RefreshToken() 
-                { 
-                    Id = Helper.GetHash(refreshTokenId),
-                    ClientId = clientid, 
-                    Subject = context.Ticket.Identity.Name,
-                    IssuedUtc = DateTime.UtcNow,
-                    ExpiresUtc = DateTime.UtcNow.AddMinutes(Convert.ToDouble(refreshTokenLifeTime)) 
-                };
-
-                context.Ticket.Properties.IssuedUtc = token.IssuedUtc;
-                context.Ticket.Properties.ExpiresUtc = token.ExpiresUtc;
-                
-                token.ProtectedTicket = context.SerializeTicket();
-
-                var result = await _repo.AddRefreshToken(token);
-
-                if (result)
+                CleanupRefreshTokens(_repo, token);
+                if (_repo.AddRefreshToken(token))
                 {
                     context.SetToken(refreshTokenId);
                 }
-             
             }
         }
 
-        public async Task ReceiveAsync(AuthenticationTokenReceiveContext context)
+        private static string GetClientId(BaseContext context)
         {
+            return context.OwinContext.Get<string>(Constants.OAuth.ClientId);
+        }
 
-            var allowedOrigin = context.OwinContext.Get<string>("as:clientAllowedOrigin");
-            context.OwinContext.Response.Headers.Add("Access-Control-Allow-Origin", new[] { allowedOrigin });
+        private static string GetUserAgent(IOwinContext context)
+        {
+            return context.Request.Headers.Get("User-Agent").Substring(0, 50);
+        }
 
-            string hashedTokenId = Helper.GetHash(context.Token);
+        public void Receive(AuthenticationTokenReceiveContext context)
+        {
+            var hashedTokenId = Helper.GetHash(context.Token);
 
-            using (AuthRepository _repo = new AuthRepository())
+            using (var _repo = new AuthRepository())
             {
-                var refreshToken = await _repo.FindRefreshToken(hashedTokenId);
+                var refreshToken = _repo.FindRefreshToken(hashedTokenId);
 
-                if (refreshToken != null )
+                if (refreshToken == null)
                 {
-                    //Get protectedTicket from refreshToken class
-                    context.DeserializeTicket(refreshToken.ProtectedTicket);
-                    var result = await _repo.RemoveRefreshToken(hashedTokenId);
+                    // Refresh token is not valid
+                    // TODO: log suspicious activity
+                    return;
                 }
+                if (refreshToken.ClientId != GetClientId(context))
+                {
+                    // Refresh token is issued to a different clientId.
+                    // TODO: log suspicious activity
+                    return;
+                }
+                var user = _repo.FindUserByName(refreshToken.UserName);
```

Comment:
I'm trying to understand this one. You load a refresh token by the hashed token id. Then you load the user by the user name of this token. What is the benefit of this approach? The comment "Refresh token was issued for the other user" doesn't make sense to me. Can you pls explain what your intention was? Thx!

Reply:
Hi @VILLAN3LL3 ,
Well it's been very long time since I looked in this code so can guarantee that everything is correct. From what I see and as far as I remember refresh token hash is the only thing which is required to authorize the user and issue a fresh access token.
So the only way to know for which user to issue a new access token is to keep user name associated to refresh token. So the code in line 78 looks up the user in the database and the check in the line 79 is redundant if you never delete users and I guess I put it there just to avoid resharper warning below.
So bottom line the comment below is misleading. If we actually end up in the line 83 it means that the user, for whom the token was issued - now doesn't exist in the database.
Hope that helps